Episcope - Integrative ATAC-RNA Framework for Condition-Specific Gene Regulatory Networks
===========================================================================================

[![Version](https://img.shields.io/badge/version-1.0.0-blue.svg?style=plastic)](https://github.com/oncologylab/episcope)
[![License](https://img.shields.io/badge/license-MIT-green.svg?style=plastic)](https://github.com/oncologylab/episcope/blob/main/LICENSE)
[![Bioconductor](https://img.shields.io/badge/install%20via-BiocManager-orange.svg?style=plastic)](https://bioconductor.org)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=plastic)](https://github.com/oncologylab/episcope/graphs/commit-activity)
[![publication](https://img.shields.io/badge/Publication-in%20prep-blue.svg?style=plastic)]()

---

## Introduction

**Episcope** is a modular framework that integrates **chromatin accessibility (ATAC)** and **transcriptomic (RNA-seq)** data to infer condition-specific **TF binding sites (TFBS)**, and construct dynamic **gene regulatory networks (GRNs)**.

It enables users to:
 
- Collapse redundant ATAC footprints (FPs) generated by standard pipelines into **consensus, site- and motif-nonredundant footprint clusters** based on genomic overlap and shared motif composition
- Infer **condition-specific** canonical and non-canonical TFBS by correlating changes in TF expression with footprint accessibility scores across conditions 
- Refine **regulatory priors** (e.g., GeneHancer, genomic proximity, or user-supplied 3D chromatin interaction data) by filtering **TFâ†’FPâ†’gene** links using concordant correlation between *(i)* TF expression and target expression and *(ii)* footprint accessibility and target expression  
- Extract active **TFâ†’FPâ†’gene regulatory links** within each condition and identify **differential** links between conditions 
- Model dynamic GRNs using **variational autoencoder (VAE)**-based topic modeling to uncover recurrent and transitional regulatory states
- Generate **interactive** visualizations of **topic- and condition-specific GRNs**, along with cross-topic/condition summary plots of pathway enrichment, TF-target clustering, and top master TFs

---

## Installation

Episcope is an R package available on GitHub.

```r
# Using remotes
remotes::install_github("oncologylab/episcope")

# or using pak
pak::pak("oncologylab/episcope")
```

Dependencies (Bioconductor and CRAN):
```r
install.packages(c("visNetwork", "igraph", "ggplot2", "data.table", "BiocManager"))
BiocManager::install(c("DESeq2", "GenomicRanges", "SummarizedExperiment"))
```

---

## Usage Overview

Episcope runs as a 3-module pipeline and each module contains 2-3 primary functions and optional auxiliary functions:

### [**Module 1: Predict TF binding sites (TFBS)**](https://github.com/oncologylab/episcope/wiki/Step-1-Predict-TF-binding-sites)

#### **Primary functions**
- [`load_multiomic_data()`](link) produces a combined multiomic data object from a `*.yaml*` configuration file specifying: *(i)* paths to the merged, normalized ATAC master table and normalized RNA-seq master table, *(ii)* the directory containing per-condition TF motif footprint score files generated by TOBIAS or fp-tools, and *(iii)* a metadata table mapping samples to conditions, along with other configuration settings.
- [`correlate_tf_to_fp()`](link) first aligns and filters redundant TF footprints (FPs) to compute consensus TF clusters, and performs quantile normalization on the collapsed, non-redundant FPs. It then computes Pearson (linear) and Spearman (rank-based) correlations between TF expression and footprint scores across conditions using the combined multi-omic data object. In *canonical* mode, each TF is correlated only with footprints matching its canonical binding motif(s), whereas in *all* mode, TF expression is correlated with all footprint scores.

#### **Auxiliary functions**
- [`output_per_condition_tfbs_bed_files()`](link)
- [`experimental_benchmarking_of_tfbs_predictions()`](link)

### [**Module 2: Connect TFs to target genes**](https://github.com/oncologylab/episcope/wiki/Step-2-Connect-TFs-to-Target-Genes)

#### **Primary functions**
- [`link_tf_targets()`](link) links TF binding sites (TFBS; predicted in Module 1 or user-supplied) to candidate target genes using GeneHancer enhancerâ€“gene mappings (default), or user-defined genomic distance constraints (e.g., 30kb), or user-supplied 3D chromatin interaction loop file (e.g., Hi-C), It then computes cross-condition *(i)* Pearson correlations between TF and candidate target expression and *(ii)* Pearson and Spearman correlations between footprint scores and candidate target expression, and filters TF->FP->target links that pass correlation cutoffs specified in the `*.yaml*` configuration file. 
- [`extract_active_links_by_condition()`](link) generates per-condition summary tables of active (link_status = 1) TF->FP->target links. A filtered TF->FP->target link from above is flagged as active in a given condition if both the TF and target gene are expressed above the cutoff specified in the `*.yaml*` configuration file, **and** if the corresponding footprint (FP) score and ATAC peak signal in that condition also exceed their respective cutoffs (defined in the `*.yaml*` file).

#### **Auxiliary functions**
- [`pathway_enrichment_of_tf_targets()`](link)
- [`build_custom_grns()`](link)

### [**Module 3: Learn regulatory topics & visualize differential GRNs**](https://github.com/oncologylab/episcope/wiki/Step-3-Perform-differential-GRN-analysis-&-identify-master-TFs)

#### **Primary functions**
- [`find_differential_links()`](link) performs pairwise comparisons and extracts differential TF->FP->target links based on a user-supplied comparison meta table. 
- [`train_topic_models()`](link) constructs **perâ€“TF-cluster â€œdocumentsâ€** containing condition- and direction-specific differential TFâ†’FPâ†’target links, then trains **multilayer perception (MLP)** and **variational autoencoder (VAE)**-based topic models and output model matrices for a series of user-defined k values. 
- [`extract_regulatory_topics()`](link) computes link-topic probabilities and automatically assigns links to topics based on a user-selected k, and generates topic summary plots. 

#### **Auxiliary functions**
- [`run_per_condition_topic_clustering()`](link) 
- [`run_per_condition_topic_pathway_enrichment()`](link) 
- [`build_per_condition_topic_differential_grns()`](link) 
- [`build_custom_differential_grns()`](link) 

---

## Pipelines

episcope supports multiple orchestrated pipelines:

- **Snakemake** - scalable, HPC-optimized  
- **Nextflow** - reproducible, cloud-ready  
- **R scripts** - local exploratory runs

ğŸ‘‰ https://github.com/oncologylab/episcope_snakemake  
ğŸ‘‰ https://github.com/oncologylab/episcope_nextflow

---

## Help and Documentation

- ğŸ“˜ Docs https://github.com/oncologylab/episcope/wiki  
- â“ FAQ https://github.com/oncologylab/episcope/wiki/FAQ  
- ğŸ› Bug report https://github.com/oncologylab/episcope/issues

---

## How to Cite

Li, Y., Yi, C. *et al.* (in preparation).  
**Episcope:** Integrative ATAC-RNA framework for condition-specific gene regulatory network analysis.

---

## License

This project is licensed under the [GNU General Public License v3.0](https://github.com/oncologylab/episcope/blob/main/LICENSE.md).
